src/core/pointcloud/qgspointcloudstatscalculator.cpp

```diff
@@ -0,0 +1,219 @@
+        summary.minimum = std::numeric_limits<double>::max();
+        summary.maximum = std::numeric_limits<double>::lowest();
+        summary.count = 0;
+        summary.classCount.clear();
+        statsMap[ attribute.name() ] = summary;
+      }
+
+      std::unique_ptr<QgsPointCloudBlock> block( mIndex->nodeData( node, mRequest ) );
+
+      const char *ptr = block->data();
+      int count = block->pointCount();
+      const QgsPointCloudAttributeCollection attributes = block->attributes();
+      int recordSize = attributes.pointRecordSize();
+
+      for ( int i = 0; i < count; ++i )
+      {
+        for ( QgsPointCloudAttribute attribute : attributes.attributes() )
+        {
+          double attributeValue = 0;
+          int attributeOffset = 0;
+          attributes.find( attribute.name(), attributeOffset );
+
+          QgsPointCloudStatsCalculator::AttributeStatistics &stats = statsMap[ attribute.name() ];
+          QgsPointCloudRenderContext::getAttribute( ptr, i * recordSize + attributeOffset, attribute.type(), attributeValue );
+          stats.minimum = std::min( stats.minimum, attributeValue );
+          stats.maximum = std::max( stats.maximum, attributeValue );
+          stats.count++;
+          if ( attribute.name() == QStringLiteral( "Classification" ) )
+          {
+            stats.classCount[( int )attributeValue ]++;
+          }
+        }
+      }
+      return statsMap;
+    }
+  private:
+    QgsPointCloudIndex *mIndex = nullptr;
+    QgsPointCloudRequest mRequest;
+};
+
+
+
+void QgsPointCloudStatsCalculator::calculateStats( const QVector<QgsPointCloudAttribute> &attributes, qint64 pointsLimit )
+{
+  if ( !mIndex->isValid() )
+  {
+    QgsMessageLog::logMessage( QObject::tr( "Unable to calculate statistics of an invalid index" ) );
+    return;
+  }
+  mRequest.setAttributes( attributes );
+
+  //
+  qint64 pointCount = 0;
+  QVector<IndexedPointCloudNode> nodes;
+  QQueue<IndexedPointCloudNode> queue;
+  queue.push_back( mIndex->root() );
+  while ( !queue.empty() )
+  {
+    IndexedPointCloudNode node = queue.front();
+    queue.pop_front();
+    if ( !mProcessedNodes.contains( node ) )
+      pointCount += mIndex->nodePointCount( node );
+    if ( pointsLimit != -1 && pointCount > pointsLimit )
+      break;
+    if ( !mProcessedNodes.contains( node ) )
+      nodes.push_back( node );
+    for ( IndexedPointCloudNode child : mIndex->nodeChildren( node ) )
+    {
+      queue.push_back( child );
+    }
+  }
+
+  QgsPointCloudStatsCalculationTask *task = new QgsPointCloudStatsCalculationTask( mIndex, attributes, nodes );
+  QObject::connect( task, &QgsTask::taskCompleted, this, &QgsPointCloudStatsCalculator::statsCalculationFinished );
+
+  mStatsCalculationTaskId = QgsApplication::taskManager()->addTask( task );
+}
+
+void QgsPointCloudStatsCalculator::statsCalculationFinished()
+{
+  QgsPointCloudStatsCalculationTask *task = qobject_cast<QgsPointCloudStatsCalculationTask *>( QObject::sender() );
```

Comment:
hmm, the QgsTask should not be intertwined into stats calculator... such QgsTask should be rather just a wrapper around the stats calculator class that will call it, watch its progress and cancel it if needed... The rationale is that we may not want to _always_ use task manager when doing stats calculation.

Reply:
My initial thought was to have the ability to enhance the calculated statistics by having the same object calculate stats multiple times. That's why the task is inside the stats calculation class.
Sorry but I don't see why we wouldn't be always using the task for stats calculation.